MoreLinq/Zip.g.tt

```diff
@@ -0,0 +1,379 @@
+        /// </remarks>
+        public static IEnumerable<(<#= o.Types#>)> EquiZip<<#= o.Types#>>(
+<# foreach (var arg in o.Arguments) { #>
+            <#= arg.Num == 1 ? "this " : "" #>IEnumerable<T<#= arg.Number #>> <#= arg.Ordinal #>Source<#= arg.Num < o.Count ? "," : ")" #>
+<#}#>
+        {
+            return EquiZip(
+<# foreach (var arg in o.Arguments) { #>
+                <#= arg.Ordinal #>Source,
+<#}#>
+                ValueTuple.Create);
+        }
+
+        /// <summary>
+        /// Returns a projection of tuples, where each tuple contains the N-th
+        /// element from each of the input sequences. The resulting sequence
+        /// will always be as long as the longest of input sequences where the
+        /// default value of each of the shorter sequence element types is used
+        /// for padding.
+        /// </summary>
+<# foreach (var arg in o.Arguments) { #>
+        /// <typeparam name="T<#= arg.Number #>">Type of elements in <#= arg.Ordinal #> input sequence.</typeparam>
+<#} #>
+        /// <typeparam name="TResult">Type of elements in result sequence.</typeparam>
+<# foreach (var arg in o.Arguments) { #>
+        /// <param name="<#= arg.Ordinal #>Source">The <#= arg.Ordinal #> source sequence.</param>
+<#} #>
+        /// <param name="resultSelector">
+        /// Function to apply to each tuple of elements.</param>
+        /// <returns>
+        /// A projection of tuples, where each tuple contains the N-th element
+        /// from each of the argument sequences.</returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+        public static IEnumerable<TResult> ZipLongest<<#= o.Types#>, TResult>(
+<# foreach (var arg in o.Arguments) { #>
+            <#= arg.Num == 1 ? "this " : "" #>IEnumerable<T<#= arg.Number #>> <#= arg.Ordinal #>Source,
+<#}#>
+            Func<<#= o.Types#>, TResult> resultSelector)
+        {
+<# foreach (var arg in o.Arguments) { #>
+            if (<#= arg.Ordinal #>Source == null) throw new ArgumentNullException(nameof(<#= arg.Ordinal #>Source));
+<#} #>
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            return _(); IEnumerable<TResult> _()
+            {
+<# foreach (var arg in o.Arguments) { #>
+                IEnumerator<T<#= arg.Number #>> e<#= arg.Number #> = null;
+<#} #>
+
+                try
+                {
+<# foreach (var arg in o.Arguments) { #>
+                    e<#= arg.Number #> = <#= arg.Ordinal #>Source.GetEnumerator();
+<#} #>
+
+<# foreach (var arg in o.Arguments) { #>
+                    var v<#= arg.Number #> = default(T<#= arg.Number #>);
+<#} #>
+
+                    while (
+<# foreach (var arg in o.Arguments) { #>
+                        ZipHelper.MoveNextOrDefault<T<#= arg.Number #>>(ref e<#= arg.Number #>, ref v<#= arg.Number #>)<#= arg.Num < o.Count ? " |" : ")" #>
```

Comment:
You want conditional OR (`||`) instead of the bitwise OR `|` here…

```suggestion
                        ZipHelper.MoveNextOrDefault<T<#= arg.Number #>>(ref e<#= arg.Number #>, ref v<#= arg.Number #>)<#= arg.Num < o.Count ? " ||" : ")" #>
```

It bothers me that a test did not spot this error in refactoring. It means there's coverage missing that should be fixed.

Reply:
I use the [logical OR | operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#logical-or-operator-) here and not the `||` one because I need to evaluate all the operands. (`||` will stop evaluation if it finds a true).

Reply:
I need to evaluate all the operands here, `||` will stop evaluation on the first true.
[Logical OR operator |](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#logical-or-operator-) evaluates both operands even if the left-hand operand evaluates to true.

If you put `||` the test fails.

Reply:
With the actual implementation, `ZipLongest` will consume all the sequence and dispose them as their end are reached. And that is actually tested.

Reply:
Thanks for correcting me and like the neat trick.